### PR TITLE
[17] compiler.ast.CaseStatement.ResolvedCase.intValue not initialized for Strings #1126

### PR DESCRIPTION
**Pull request by jarthana**
## What it does
Fixed https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1126

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/SwitchPatternTest.java [hunk 1]
```diff
@@ -6182,4 +6182,25 @@ public void testIssue919() {
...
 			"----------\n");
 	}
+	public void testIssue1126() {
+		runConformTest(
+				new String[] {
+					"X.java",
+					"public class X {\n"+
+					" 	static int foo(String i) {\n"+
+					"		return switch (i) {\n"+
+					"	     case \"abc\" -> 0;\n"+
```
Comment (jukzi) [inline org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/SwitchPatternTest.java:6192]:
```suggestion
					"	     case \"FB\" -> 0;\n"+
```

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/SwitchPatternTest.java [hunk 2]
```diff
@@ -6182,4 +6182,25 @@ public void testIssue919() {
...
+					"	     case \"abcd\" -> 1;\n"+
```
Comment (jukzi) [inline org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/SwitchPatternTest.java:6193]:
```suggestion
					"	     case \"Ea\" -> 1;\n"+
```

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java [hunk 3]
```diff
@@ -978,7 +978,7 @@ public void resolve(BlockScope upperScope) {
 								IntPredicate check = (idx) -> {
 									Constant c2 = this.otherConstants[idx].c;
 									if (con.typeID() == TypeIds.T_JavaLangString) {
-										return c2.stringValue().equals(con.stringValue());
+										return c.intValue() == this.otherConstants[idx].intValue();
```
Comment (jukzi) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java:981]: That shows an Error if two Strings have the same hashCode. For exampe "FB" and "Ea" gives "Duplicate case".

**Review by jukzi — CHANGES_REQUESTED** [review]

**Comment by jarthana** [comment]
Yes, we have some failures due to the same reason. Will look into this.

**Comment by jarthana** [comment]
So, it's not useful to do the int comparison. I thought of doing some optimization of falling back to String comparison when the int comparison returns true.  But I don't see much value as the Strings are probably compared for reference and interned as well. So, have to live with the fact that the intValue() is not much useful for String constants in case.

**Comment by jukzi** [comment]
But did you try the test with hash-conflicting strings? it did still fail with revert for me.

**Comment by jarthana** [comment]
> But did you try the test with hash-conflicting strings? it did still fail with revert for me.

Yes, now with us doing a simple String comparison, works fine. Will capture this in a new test case for the record.

**Comment by jarthana** [comment]
> But did you try the test with hash-conflicting strings? it did still fail with revert for me.

@jukzi can you confirm that this no longer an issue?

**Review by jukzi — APPROVED** [review]
SwitchPatternTest now passes locally for me
image

**Pull request merged**